tsdb/head.go

```diff
@@ -1382,30 +1431,59 @@ func (a *headAppender) log() error {
 			return errors.Wrap(err, "log samples")
 		}
 	}
+	if len(a.exemplars) > 0 {
+		rec = enc.Exemplars(exemplarsForEncoding(a.exemplars), buf)
+		buf = rec[:0]
+
+		if err := a.head.wal.Log(rec); err != nil {
+			return errors.Wrap(err, "log exemplars")
+		}
+	}
 	return nil
 }
 
+func exemplarsForEncoding(es []exemplarWithSeriesRef) []record.RefExemplar {
+	ret := make([]record.RefExemplar, 0, len(es))
+	for _, e := range es {
+		ret = append(ret, record.RefExemplar{
+			Ref:    e.ref,
+			T:      e.exemplar.Ts,
+			V:      e.exemplar.Value,
+			Labels: e.exemplar.Labels,
+		})
+	}
+	return ret
+}
+
 func (a *headAppender) Commit() (err error) {
 	if a.closed {
 		return ErrAppenderClosed
 	}
 	defer func() { a.closed = true }()
-	if err := a.log(); err != nil {
-		_ = a.Rollback() // Most likely the same error will happen again.
-		return errors.Wrap(err, "write to WAL")
-	}
 
-	// No errors logging to WAL, so pass the exemplars along to the in memory storage.
+	// We need to know if an exemplar is going to be accepted into the exemplar storage before we write it to the WAL.
+	// This duplicating of (most) exemplars might not be as memory efficient as we'd like, but it's more time efficient
+	// than deleting from a slice in place.
+	commitExemplars := make([]exemplarWithSeriesRef, 0, len(a.exemplars))
 	for _, e := range a.exemplars {
 		s := a.head.series.getByID(e.ref)
-		// We don't instrument exemplar appends here, all is instrumented by storage.
-		if err := a.exemplarAppender.AddExemplar(s.lset, e.exemplar); err != nil {
-			if err == storage.ErrOutOfOrderExemplar {
-				continue
-			}
-			level.Debug(a.head.logger).Log("msg", "Unknown error while adding exemplar", "err", err)
+		err := a.exemplarAppender.AddExemplar(s.lset, e.exemplar)
```

Comment:
`We must first log these exemplars in the WAL and then add it to the storage. The WAL gives the order of append hence on restart out of order exemplars will remain out of order and need not be filtered here` The bigger issue is that this would lead to remote writing exemplars that shouldn't be stored/remote written. The only real fix here is I guess to duplicate some of the exemplar validation/checks from the storage into `head.go`.

Reply:
Here's how I'd address this right now: https://github.com/cstyan/prometheus/commit/61ff496d57d12fdca243b2e318fb087fd8754fe5

It seems similar to what we do with WAL logging of samples and then appending to Head with `appendable(...)`, though I'm not sure how we should proceed.

Reply:
@codesome can you let me know if
1) this is actually a blocking issue, see my first reply in this thread for the reasoning the order of adding to the storage and then the WAL is the way it is currently
2) If so, does the commit in the above comment resolve the issue? We'd essentially be duplicating the validation logic, and should look into improvements in the future if we go that route.